web3/pm.py

```diff
@@ -245,29 +245,28 @@ def _get_package_name(self, package_id: bytes) -> str:
     @to_tuple
     def _get_all_package_ids(self) -> Iterable[Tuple[bytes]]:
         num_packages = self._num_package_ids()
-        # Logic here b/c Solidity Reference Registry implementation returns ids in reverse order
-        package_ids = [
-            self.registry.functions.getAllPackageIds(index, (index + 4)).call()[0]
-            for index in range(0, num_packages, 4)
-        ]
-        for package_id in concat([x[::-1] for x in package_ids]):
-            yield package_id
+        pointer = 0
+        while pointer < num_packages:
+            new_ids, pointer = self.registry.functions.getAllPackageIds(
+                pointer,
+                (pointer + BATCH_SIZE)
+            ).call()
+            yield from reversed(new_ids)
 
     def _get_release_id(self, package_name: str, version: str) -> bytes:
         return self.registry.functions.getReleaseId(package_name, version).call()
 
     @to_tuple
     def _get_all_release_ids(self, package_name: str) -> Iterable[Tuple[bytes]]:
         num_releases = self._num_release_ids(package_name)
-        # Logic here b/c Solidity Reference Registry implementation returns ids in reverse order
-        release_ids = [
-            self.registry.functions.getAllReleaseIds(
-                package_name, index, (index + 4)
-            ).call()[0]
-            for index in range(0, num_releases, 4)
-        ]
-        for release_id in concat([x[::-1] for x in release_ids]):
-            yield release_id
+        pointer = 0
+        while pointer < num_releases:
```

Comment:
One more thing.  This is probably fine, however, you could save yourself a headache by enforcing that `pointer` always increases on each loop iteration to ensure you cannot end up in an infinite loop.